📈 Insights QoL Part 1

connorwinston opened the pull request:
- Updated formatting/style of **Edit/Context Menu** in the top right of panels to blur the background and to work with new small sized panels.
- Updated formatting/style of **Panel Size** helper text to match the Edit/Context Menu styling and moved it to the bottom right corner.
- Added **Comments** to insights dashboards.
- Added additional styling options to **Label and Metric** panels including, Text Wraping, Text Alignment, Font Weight, Font Style, Font Size (+ Auto Sizing), and Font
- Added new Metric List Panel which lets you display a list of metrics.
- Added new Number Formatting options to Metric and MetricList including, Style (Unit, Decimal, Currency, Percent), Notation (Standard, Scientific, Engineering, Compact), Unit, Minimum Decimals, Maximum Decimals.
- Fixed Sizing for Small Meters
- **Line Chart Enhancements** 
  - Making a line chart smaller than current minimum will make a SparkLine!
  - Fixed Naming/Logic of X-Axis / Y-Axis field selection, aggregation.
  - Added Grouping support for _Multiple Series_
  - Added Y-Axis Min/Max config options
  - Added Fill Type config option matching TimeSeries
  - Added Show Legend Option

![CleanShot 2023-08-17 at 17 57 00](https://github.com/directus/directus/assets/12714889/6b6622b5-03b0-48d4-85db-3e63d58364e4)

connorwinston committed:
Start adding Create Default Modal

connorwinston committed:
Add Metric List

connorwinston committed:
Merge branch 'main' into connor/insights-qol-upgrade

connorwinston committed:
Merge branch 'main' into connor/insights-qol-upgrade

connorwinston committed:
Number Formatter Utility

connorwinston committed:
Start on Label Formatting Options and Resizing

connorwinston committed:
Update Metric List to use New Number Formatter

connorwinston committed:
fix auto fit and styling

connorwinston committed:
fix typings for number formatter and error catching

connorwinston committed:
add conditional formatting to metric list

connorwinston committed:
Add Number Formatting to Metrics and Size Formats

connorwinston committed:
Add Option to Hide Percentage and Fix Sizing of Meter

connorwinston committed:
Add manual sizing for labels

connorwinston committed:
Add translation strings

connorwinston committed:
Merge branch 'main' into connor/insights-qol-upgrade

connorwinston committed:
Add comments to insights

connorwinston committed:
Fix color errors and add icon

connorwinston committed:
update translations

connorwinston committed:
Refine Linechart

connorwinston committed:
Revert "Start adding Create Default Modal"

This reverts commit 184dd0bb73211aa816623b56258c2892926c0c9c.

connorwinston committed:
Run prettier

connorwinston committed:
Remove Console Logging

connorwinston committed:
Fix Typings

connorwinston committed:
Create itchy-berries-rule.md

connorwinston committed:
Merge branch 'main' into connor/insights-qol-upgrade

connorwinston committed:
Merge branch 'connor/insights-qol-upgrade' of https://github.com/directus/directus into connor/insights-qol-upgrade

paescuj committed:
Merge branch 'next' into connor/insights-qol-upgrade

paescuj committed:
Format file

connorwinston committed:
Fix Metric List for Aggre Group PR

connorwinston committed:
Remove Empty Class

connorwinston committed:
Merge branch 'next' into connor/insights-qol-upgrade

connorwinston committed:
Merge remote-tracking branch 'origin/next' into connor/insights-qol-upgrade

rijkvanzanten committed:
Merge branch 'next' into connor/insights-qol-upgrade

rijkvanzanten committed:
Run formatter

rijkvanzanten committed:
Fix linter errors

rijkvanzanten committed:
Merge branch 'next' into connor/insights-qol-upgrade

rijkvanzanten committed:
Fix import

rijkvanzanten commented on app/src/panels/label/panel-label.vue:
```diff
@@ -1,16 +1,99 @@
+	if (props.fontSize == 'auto') {
+		updateFit();
+
+		// Delay the initial font size adjustment to allow the text/font to render fully
+		setTimeout(() => {
+			updateFit();
+		}, 500);
```
This feels brittle to me. Any way we can make this event driven?

rijkvanzanten replied:
Are we just waiting for Vue's next tick? In that case it might be worth awaiting nextTick instead

ConnorSimply replied:
This was added in because during my testing, I could not get it to consistently do the initial updateFit on load/mount/render of the text. The text needs to be fully rendered, wrapped, and displayed for updateFit to know how much to shrink/grow to make it fit right because it compares the parent's height to the child's height and shrinks it until it fits just right.

rijkvanzanten replied:
@br41nslug Could you take a look?

br41nslug replied:
@ConnorSimply can you help me reproduce the issue you mentioned running into during testing? In my tests so far with the label panel i cannot see a noticeable difference between with this `setTimeout` and without (commented out) 🤔 Tested in both Firefox and Chrome with and without throtling

paescuj replied:
IMO, this should be enough:
```ts
function updateFit() {
	adjustPadding();
	adjustFontSize();
}

onMounted(() => {
	if (props.fontSize === 'auto') {
		const container = labelContainer.value;
		if (!container) return;

		// Create a ResizeObserver to watch for changes in the container's dimensions
		resizeObserver = new ResizeObserver(() => {
			updateFit();
		});

		resizeObserver.observe(container);
	}
});

onUpdated(() => {
	if (props.fontSize === 'auto') {
		updateFit();
	}
});
```

Note: According to https://stackoverflow.com/a/60026394/20504834 `ResizeObserver` fires initially, which matches my observation in Firefox.

ConnorSimply replied:
@br41nslug Yes, you can reproduce the issue by putting like 1-2 paragraphs in the label for the text and setting the font to serif. When the font is not san-serif the text will resize successfully but it resizes BEFORE the font loads so the size that it resizes to is for sans-serif not serif.

@paescuj Yes, resizeObserver does fire initially. That is not the issue, the issue is that it fires too early before the font finishes loading. My testing locally has the font loading in roughly 200ms after the panel data loads so I figured that 500ms would be a reasonable amount of time to delay a "double check" resize incase fonts are not finished loading.

paescuj replied:
I see! So could we rely on  the https://developer.mozilla.org/en-US/docs/Web/API/FontFaceSet/ready event?

ConnorSimply replied:
That did the trick! :)

br41nslug replied:
> Yes, you can reproduce the issue by putting like 1-2 paragraphs in the label for the text and setting the font to serif. When the font is not san-serif the text will resize successfully but it resizes BEFORE the font loads so the size that it resizes to is for sans-serif not serif.

Seems it was fixed with paescuj's changes as i am unable to reproduce the issue on the current state of the branch

rijkvanzanten commented on app/src/panels/metric/panel-metric.vue:
```diff
@@ -1,39 +1,123 @@
+		updateFit();
+
+		// Delay the initial font size adjustment to allow the text/font to render fully
+		setTimeout(() => {
+			updateFit();
+		}, 500);
```
Same here

rijkvanzanten reviewed (approved):
Code overall looking really good! One quick question around a seemingly arbitrary timeout but otherwise LGTM

paescuj commented on app/src/composables/use-auto-fit-text.ts:
```diff
@@ -0,0 +1,96 @@
+
+	switch (paddingValues.length) {
+		case 1:
+			// @ts-expect-error
+			top = right = bottom = left = paddingValues[0];
+			break;
+		case 2:
+			// @ts-expect-error
+			[top, right] = paddingValues;
+			bottom = top;
+			left = right;
+			break;
+		case 3:
+			// @ts-expect-error
+			[top, right, bottom] = paddingValues;
+			left = right;
+			break;
+		case 4:
+			// @ts-expect-error
+			[top, right, bottom, left] = paddingValues;
+			break;
```
We should get rid of these "@ts-expect-error" comments, preferably by proper checks or using non-null assertion if not possible otherwise.

paescuj commented on app/src/panels/label/index.ts:
```diff
@@ -7,29 +7,237 @@ export default definePanel({
-	minWidth: 8,
-	minHeight: 6,
+	options: ({ options }) => {
+		if (!options) options = {};
```
Is this still necessary?

paescuj commented on app/src/panels/label/panel-label.vue:
```diff
@@ -1,16 +1,99 @@
+}
+
+onMounted(() => {
+	if (props.fontSize == 'auto') {
```
Strict equality operator is generally preferred
```suggestion
	if (props.fontSize === 'auto') {
```

paescuj commented on app/src/panels/label/panel-label.vue:
```diff
@@ -22,24 +105,41 @@ export default defineComponent({
 	</div>
 </template>
 
 <style scoped>
```
```
▲ [WARNING] CSS nesting syntax is not supported in the configured target environment ("chrome87", "edge88", "es2020", "firefox78", "safari14") [unsupported-css-nesting]

    <stdin>:9595:0:
      9595 │ &.sans-serif[data-v-073cd843] {
           ╵ ^
```

Unfortunately, not quite there 😢: https://caniuse.com/css-nesting

```suggestion
<style lang="scss" scoped>
```

rijkvanzanten committed:
Update app/src/panels/label/panel-label.vue

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

paescuj commented on app/src/panels/line-chart/panel-line-chart.vue:
```diff
@@ -35,6 +42,9 @@ const props = withDefaults(
 		decimals: 2,
 		color: 'var(--primary)',
+		fillType: 'gradient',
+		min: undefined,
+		max: undefined,
```
Is already undefined by default

```suggestion
```

paescuj commented on app/src/panels/metric-list/panel-metric-list.vue:
```diff
@@ -0,0 +1,227 @@
+.metric-list-item {
+	height: 36px;
+	border-bottom: var(--border-width) solid var(--border-subdued);
+	/* display: block; */
```
? 😇 
```suggestion
```

paescuj reviewed (commented):
Is this change backwards-compatible or does it need migrations / manual migration instructions / breaking change announcements?

Asking after seeing removed options like https://github.com/directus/directus/pull/19467/files#diff-e034470a9d4e782d386eb500fcc21471647ed1fe790ecd9be86ce482c92472f6L200

rijkvanzanten commented:
That's a question better suited for @ConnorSimply to answer 🙂

connorwinston committed:
Update app/src/panels/label/panel-label.vue

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

connorwinston committed:
Update app/src/panels/line-chart/panel-line-chart.vue

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

connorwinston committed:
Remove commented css

connorwinston committed:
Remove unneeded logic

connorwinston committed:
Fix auto resizing and text alignment

connorwinston committed:
Remove Expect Error Typechecking

connorwinston committed:
Add Migration

ConnorSimply commented:
> Is this change backwards-compatible or does it need migrations / manual migration instructions / breaking change announcements?
> 
> Asking after seeing removed options like https://github.com/directus/directus/pull/19467/files#diff-e034470a9d4e782d386eb500fcc21471647ed1fe790ecd9be86ce482c92472f6L200

The only migration that needs to be made is for the metric panel to convert `abbreviate` and `decimals` to their new formatting options.

The changes to line chart are breaking and will require reconfiguration as needed by users.

connorwinston committed:
Fix Prettier

paescuj committed:
Lint auto-fixes

paescuj commented on app/src/lang/translations/en-US.yaml:
```diff
@@ -2291,3 +2293,145 @@ operations:
     batch: Batch
     batch_size: Batch Size
     cache: Cache
+core_unit_identifier: Core Unit Identifier
```
Wondering if we should nest those values, e.g. all units under `units` and others under a corresponding panel key 🤔

paescuj commented on app/src/panels/label/panel-label.vue:
```diff
@@ -1,16 +1,92 @@
+	return;
+}
+
+async function updateFit() {
+	if (props.fontSize !== 'auto' || props.text.length <= 0) return;
+	await document.fonts.ready;
+	adjustPadding();
+	adjustFontSize();
+
+	if (!resizeObserver) {
+		const container = labelContainer.value;
+		if (!container) return;
+
+		// Create a ResizeObserver to watch for changes in the container's dimensions
+		resizeObserver = new ResizeObserver(() => {
+			updateFit();
+		});
+
+		resizeObserver.observe(container);
+	}
+
+	adjustFontSize();
+}
+
+onMounted(() => {
+	updateFit();
+});
```
- I would have guessed that we only need to wait for the fonts to be ready once on mount
- Let's create the observer on mount instead of checking on each `updateFit` call

```suggestion
function updateFit() {
	if (props.fontSize !== 'auto' || props.text.length <= 0) return;

	adjustPadding();
	adjustFontSize();
}

onMounted(async () => {
	const container = labelContainer.value;
	if (!container) return;

	await document.fonts.ready;

	// Create a ResizeObserver to watch for changes in the container's dimensions
	resizeObserver = new ResizeObserver(() => {
		updateFit();
	});

	resizeObserver.observe(container);
});
```

ConnorSimply replied:
We should only instantiate a resize observer for panels that need it, which is only panels with the font size set to auto. Your suggestion creates a resize observer that observes every label and metric panel regardless of their config. This would be a waste of browser resources. By including that logic in the updateFit function we can ensure that we check the necessary conditions every time it is mounted/updated and then only deploy a resize observer if needed. I have added additional logic to cleanup resize observers when if a panel no longer needs it but is still mounted.

paescuj commented on app/src/panels/metric-list/index.ts:
```diff
@@ -0,0 +1,422 @@
+	id: 'metric-list',
+	name: 'Metric List',
+	icon: 'format_list_numbered_rtl',
+	description: 'List and visualize aggregate metrics from a collection',
```
Missing translation

connorwinston committed:
Fix Resizing CSS and JS

connorwinston committed:
Merge branch 'connor/insights-qol-upgrade' of https://github.com/directus/directus into connor/insights-qol-upgrade

connorwinston committed:
Add Metric List Translation

paescuj committed:
Optimize migration

paescuj committed:
Format files

paescuj committed:
Fix metric-list translations

paescuj committed:
Fix type issue in metric-list/index.ts

> Default export of the module has or is using private name 'Aggregate'.ts(4082)
> Default export of the module has or is using private name 'Group'.ts(4082)

paescuj commented on app/src/lang/translations/en-US.yaml:
```diff
@@ -2291,3 +2296,145 @@ operations:
     batch: Batch
     batch_size: Batch Size
     cache: Cache
+core_unit_identifier: Core Unit Identifier
```
On a second thought: I'd really like to have the units translations nested under a key to make the affiliation clear.
Also I'm not sure whether some of these added translations are really used. For example I cannot see where `core_unit_identifier` is used. Can we clean those up if unused?

ConnorSimply replied:
I just realized that none of the units are needed, at the beginning the unit was a dropdown instead of a text input but now that it is a text input these are no longer needed. They have been removed.

paescuj replied:
Ah okay, I see 👍 Thanks for the clean-up ❤️

connorwinston committed:
Remove Unit Translations

connorwinston committed:
Merge remote-tracking branch 'origin/next' into connor/insights-qol-upgrade

rijkvanzanten committed:
Merge branch 'next' into connor/insights-qol-upgrade

rijkvanzanten committed:
De-duplicate migration
